### PR TITLE
New action widget to manipulate tags

### PR DESCRIPTION
**Pull request by matabele**
ActionMangleTags Widget
-----------------------------------

Whilst updating my button widgets it occurred to me that my new ActionMangleTags widget might be appropriate for the core. Just as the ActionSetField widget has largely superseded the need for the FieldMangler widget for manipulating fields, the ActionMangleTags widget would largely supersede the need for the FieldMangler widget for manipulating tags.

Action widgets are generally more straightforward in usage than widgets relying on widget messaging -- and I feel that the ActionMangleTags widget would be a useful addition to the set.

The widget is based upon my old stackable MangleTags widget and can be used to unconditionally add or remove a tag from a target tiddler, or may be used to conditionally search for and replace a tag. I have included documentation to demonstrate the use of the widget.

My original widget did not employ the $ prefix for attribute names -- I have added the prefix to bring the syntax in line with that of the other action widgets (although this is unnecessary.)

I have also signed the contributors licence, should the widget be found suitable for the core.

### CODE REVIEW

**Comment by Jermolene** [comment]
Hi @matabele great stuff, thank you for sharing.

It would be useful to extend it with a `$field` attribute to allow a field other than the tags field to be specified (the `list` field is useful). I'll make a couple more comments on the code itself.

Finally, another bit of bureaucracy: would you mind putting your CLA signature in a separate pull request? That way I can merge it immediately.

File: core/modules/widgets/action-mangletags.js [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  /*
+  Invoke the action associated with this widget
+  */
+  MangleTagsWidget.prototype.invokeAction = function (triggeringWidget, event) {
+    // Get the target tiddler //
+    var tiddler = this.wiki.getTiddler(this.targetTiddler);
+    var tag = (this.addTag || "").trim();
+    var modification = this.wiki.getModificationFields();
+    modification.tags = (tiddler.fields.tags || []).slice(0);
```
Comment (Jermolene) [inline core/modules/widgets/action-mangletags.js]: When reading the list field, use `$tw.utils.parseStringArray()` to coerce the value into an array; that will enable it to be used with fields that haven't been predeclared as containing a list-type value (ie "tags" and "list")

File: core/modules/widgets/action-mangletags.js [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+          $tw.utils.pushTop(modification.tags, tag);
+        }
+      }
+      // Else if there is an add= attribute only -- unconditionally add the tag
+    } else {
+      if (tag !== "") {
+        $tw.utils.pushTop(modification.tags, tag);
+      }
+    }
+    this.wiki.addTiddler(new $tw.Tiddler(tiddler, modification));
```
Comment (Jermolene) [inline core/modules/widgets/action-mangletags.js]: Similarly, here, you should write the list as a string via `$tw.utils.stringifyList()`, again to allow string fields to be used.

File: editions/tw5.com/tiddlers/widgets/ActionMangleTagsWidget.tid [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+* find and replace a tag in the tags field
+
+! Content and Attributes
+
+|Attribute |Explanation |h
+|$tiddler |Title of the tiddler to manipulate (if not provided defaults to the [[current tiddler|Current Tiddler]] |
+|$add |Optional name of a tag to append to the tags field |
+|$remove |Optional name of a tag to find and remove |
+
+! Examples
```
Comment (Jermolene) [inline editions/tw5.com/tiddlers/widgets/ActionMangleTagsWidget.tid]: You'll see that the other widget documentation separates the examples out into separately transcluded tiddlers. Would you mind following the same style?

**Comment by Jermolene** [comment]
Thanks again, and welcome to the core!

File: core/modules/widgets/action-mangletags.js (hunk 1 above)
Comment (matabele) [inline core/modules/widgets/action-mangletags.js]: Don't know why this didn't occur to me -- I needed a widget to do exactly this some time ago.

File: editions/tw5.com/tiddlers/widgets/ActionMangleTagsWidget.tid (hunk 3 above)
Comment (matabele) [inline editions/tw5.com/tiddlers/widgets/ActionMangleTagsWidget.tid]: Will do.

**Pull request closed without merging**